### PR TITLE
Fixed the callbacks based on the routing spike

### PR DESCRIPTION
**Pull request by SzymonPobiega**
Based on https://github.com/Particular/NServiceBus/pull/3374

 * Fixed `TaskCompletionSourceAdapter` to properly handle cancelled tasks
 * Moved behavior classes to proper folders
 * Adjusted the scale out sample

### CODE REVIEW

**Comment by SzymonPobiega** [comment]
@Particular/nservicebus-maintainers please review.

File: src/NServiceBus.Callbacks.AcceptanceTests/NServiceBus.Callbacks.AcceptanceTests.csproj [hunk 1]
```diff
@@ -29,13 +29,19 @@
     <ErrorReport>prompt</ErrorReport>
     <WarningLevel>4</WarningLevel>
   </PropertyGroup>
+  <PropertyGroup>
+    <SignAssembly>true</SignAssembly>
+  </PropertyGroup>
+  <PropertyGroup>
+    <AssemblyOriginatorKeyFile>NServiceBus.snk</AssemblyOriginatorKeyFile>
```
Comment (andreasohlund) [inline src/NServiceBus.Callbacks.AcceptanceTests/NServiceBus.Callbacks.AcceptanceTests.csproj]: Is it needed to strong name the acceptance tests?

File: src/NServiceBus.Callbacks/InternalsVisibleTo.cs [hunk 2]
```diff
@@ -0,0 +1,4 @@
+﻿using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("NServiceBus.Callbacks.AcceptanceTests, PublicKey=0024000004800000940000000602000000240000525341310004000001000100dde965e6172e019ac82c2639ffe494dd2e7dd16347c34762a05732b492e110f2e4e2e1b5ef2d85c848ccfb671ee20a47c8d1376276708dc30a90ff1121b647ba3b7259a6bc383b2034938ef0e275b58b920375ac605076178123693c6c4f1331661a62eba28c249386855637780e3ff5f23a6d854700eaa6803ef48907513b92")]
```
Comment (andreasohlund) [inline src/NServiceBus.Callbacks/InternalsVisibleTo.cs]: This feels like a smell, what do you need to access in the callbacks assembly?

File: src/NServiceBus.Callbacks.AcceptanceTests/NServiceBus.Callbacks.AcceptanceTests.csproj (hunk 1 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Callbacks.AcceptanceTests/NServiceBus.Callbacks.AcceptanceTests.csproj]: `InternalsVisitbleTo` :\

File: src/NServiceBus.Callbacks/TaskCompletionSourceAdapter.cs [hunk 3]
```diff
@@ -15,7 +15,7 @@ public TaskCompletionSourceAdapter(object taskCompletionSource)
 
         public void SetResult(object result)
         {
-            var method = taskCompletionSource.GetType().GetMethod("SetResult");
+            var method = taskCompletionSource.GetType().GetMethod("TrySetResult");
```
Comment (danielmarbach) [inline src/NServiceBus.Callbacks/TaskCompletionSourceAdapter.cs]: Why did you change that?

File: src/NServiceBus.Callbacks.AcceptanceTests/NServiceBus.Callbacks.AcceptanceTests.csproj (hunk 1 above)
Comment (danielmarbach) [inline src/NServiceBus.Callbacks.AcceptanceTests/NServiceBus.Callbacks.AcceptanceTests.csproj]: But why? Feels like a smell that ATTs need to know internals

File: src/NServiceBus.Callbacks/InternalsVisibleTo.cs (hunk 2 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Callbacks/InternalsVisibleTo.cs]: Because `UniqueInstanceIdProvider` is internal and I placed its test in the acceptance tests (it plays with files a bit).

File: src/NServiceBus.Callbacks/TaskCompletionSourceAdapter.cs (hunk 3 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Callbacks/TaskCompletionSourceAdapter.cs]: Because that works even if a task is cancelled. In `When_using_callback_to_get_message_canceled` if you would disable the `done` condition you would see that calling `SetResult` results in an exception. It was masked because `done` kicked in first and was killing the endpoint. 

I used `TrySetResult` as an equivalent of setting the result if the task is not cancelled.

File: src/NServiceBus.Callbacks.AcceptanceTests/NServiceBus.Callbacks.AcceptanceTests.csproj (hunk 1 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Callbacks.AcceptanceTests/NServiceBus.Callbacks.AcceptanceTests.csproj]: I can move that test to the other test project, no problem with that. It just felt like an acceptance kind of test.

File: src/NServiceBus.Callbacks/TaskCompletionSourceAdapter.cs (hunk 3 above)
Comment (danielmarbach) [inline src/NServiceBus.Callbacks/TaskCompletionSourceAdapter.cs]: Wouldn't it make sense to change the whole adapter then as well?

Comment (SzymonPobiega) [inline src/NServiceBus.Callbacks/TaskCompletionSourceAdapter.cs]: Not sure. As I can see now, `SetException` is never called and I did not prove that thaere is something wrong with `SetCancelled`.

**Comment by SzymonPobiega** [comment]
I moved the offending test to the unit tests and switched `InternalsVisibleTo` to point to unit tests.

File: src/NServiceBus.Callbacks/TaskCompletionSourceAdapter.cs (hunk 3 above)
Comment (danielmarbach) [inline src/NServiceBus.Callbacks/TaskCompletionSourceAdapter.cs]: My point was more to rename the outer method to TrySetResult as well

Comment (SzymonPobiega) [inline src/NServiceBus.Callbacks/TaskCompletionSourceAdapter.cs]: :+1:

File: src/NServiceBus.Callbacks/UniqueInstanceIdProvider.cs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    class UniqueInstanceIdProvider : IWantToRunBeforeConfigurationIsFinalized
+    {
+        public void Run(SettingsHolder settings)
+        {
+            settings.RegisterGenerator(() => EnsureFileExists(settings), EndpointInstanceGenerationScope.Plugin, "NServiceBus.Callbacks");
+        }
+
+        internal static string EnsureFileExists(SettingsHolder settings)
+        {
+            var baseDirectory = AppDomain.CurrentDomain.BaseDirectory;
```
Comment (johnsimons) [inline src/NServiceBus.Callbacks/UniqueInstanceIdProvider.cs]: Not sure it is safe to assume the running user has writes access to the installed folder

File: src/NServiceBus.Callbacks.AcceptanceTests/When_a_callback_for_local_message_canceled.cs [hunk 5]
```diff
@@ -23,7 +23,7 @@ await Scenario.Define<Context>()
 
                             var options = new SendOptions();
 
-                            options.RouteToLocalEndpointInstance();
+                            options.RouteToThisEndpoint();
```
Comment (johnsimons) [inline src/NServiceBus.Callbacks.AcceptanceTests/When_a_callback_for_local_message_canceled.cs]: Another option ? What is this for ?

Comment (SzymonPobiega) [inline src/NServiceBus.Callbacks.AcceptanceTests/When_a_callback_for_local_message_canceled.cs]: this is `SendLocal`. We have two options for sending local: `RouteToThisEndpoint` and `RouteToThisInstance`

File: src/NServiceBus.Callbacks/UniqueInstanceIdProvider.cs (hunk 4 above)
Comment (SzymonPobiega) [inline src/NServiceBus.Callbacks/UniqueInstanceIdProvider.cs]: It is not. I had a conversation about that with @andreasohlund and we might come with different strategies here (local storage?). This PR is meant to show how it might work. The next question would be: how and when we auto-generate the individual queue names.

Comment (andreasohlund) [inline src/NServiceBus.Callbacks/UniqueInstanceIdProvider.cs]: :+1: this wouldn't work for azure as the disks are gone if the vm is moved so we would end up creating orphan queues that the user would have to clean up manually. For those users the "role instance id" would be more appropriate. cc @Particular/azure-maintainers 

I had the crazy idea to have the callbacks force the user to provide the unique id since they are the only ones that can know which ones to pick. Then our hosts can provide good defaults as needed.

Comment (johnsimons) [inline src/NServiceBus.Callbacks/UniqueInstanceIdProvider.cs]: @andreasohlund there is no need for it, the id can be random

Comment (johnsimons) [inline src/NServiceBus.Callbacks/UniqueInstanceIdProvider.cs]: This is actually the way @SeanFeldman is also doing it in Azure impl, they just created a comcept of a session id.

Comment (andreasohlund) [inline src/NServiceBus.Callbacks/UniqueInstanceIdProvider.cs]: >there is no need for it, the id can be random


So each time the endpoint starts we get a new queue? (is it up to the users to clean those queues up unless the transport has a concept of temporary queues?)

Comment (johnsimons) [inline src/NServiceBus.Callbacks/UniqueInstanceIdProvider.cs]: Kind of, at install time we create a random queue for callbacks, and then we use that as the callbacks queue at runtime (hence the reason we need to store the random name somewhere).
So everytime the user installs the endpoint a new random queue is created.

Does that make sense @andreasohlund ?

And yes, for rabbit those queues can be made non durable temporary queues, so no need to clean up at all, and they are auto created on every run.

Comment (johnsimons) [inline src/NServiceBus.Callbacks/UniqueInstanceIdProvider.cs]: Also, in Azure, I think this "temp" queues can be decomissioned automatically.

Comment (andreasohlund) [inline src/NServiceBus.Callbacks/UniqueInstanceIdProvider.cs]: My point was that if we force the users to give it to us we don't have to worry about where to store it since on azure we can't use the local disk?

And if they give it to us they can choose (and our hosts can have a good default) something that makes sense for their environment so instead of a "guid" they will see a queue that has some kind of meaning and I think that OPS people would appreciate that?

Comment (johnsimons) [inline src/NServiceBus.Callbacks/UniqueInstanceIdProvider.cs]: The problem is that majority will hardcode a value without any consideration, and once they scale-out, callbacks break.
The random naming of the queue can be something simple, eg `callbacks-[unquenumber]`

Comment (johnsimons) [inline src/NServiceBus.Callbacks/UniqueInstanceIdProvider.cs]: This way we are safe by default when scaling out, no need for the users to do anything at all.
Also, this also works for side by side upgrades.

Comment (andreasohlund) [inline src/NServiceBus.Callbacks/UniqueInstanceIdProvider.cs]: >This way we are safe by default when scaling out, no need for the users to do anything at all.

Unless the user is on azure then he has to change strategy or create some script to remove the orphan queues? ( @yvesgoeleven is there some concept of temporary queues that we can use)

Regardless, if he runs on azure against a sqlazure there would definitely be orphan queues?

Comment (danielmarbach) [inline src/NServiceBus.Callbacks/UniqueInstanceIdProvider.cs]: Have we considered the Ops part of that decision as well?

Comment (SzymonPobiega) [inline src/NServiceBus.Callbacks/UniqueInstanceIdProvider.cs]: Should I move this discussion to a separate issue? I think it is worth it (transparency etc.)

Comment (SzymonPobiega) [inline src/NServiceBus.Callbacks/UniqueInstanceIdProvider.cs]: Moving the discussion to https://github.com/Particular/NServiceBus/issues/3406

Comment (SeanFeldman) [inline src/NServiceBus.Callbacks/UniqueInstanceIdProvider.cs]: @andreasohlund @johnsimons Azure can have queues that are auto-deleted. This would require to to mark those appropriately when created. Be aware that we're assuming that application has managing rights. Some customers can run endpoints without those rights (can't create queues).

One comment about "sessions", in order to use sessions, queue has to be flagged appropriately and messages sent to that queue can be only messages with a session set.

**Comment by SzymonPobiega** [comment]
@Particular/nservicebus-maintainers please re-review as https://github.com/Particular/NServiceBus/pull/3374 has already been merged to develop.

File: src/NServiceBus.Callbacks/TaskCompletionSourceAdapter.cs [hunk 6]
```diff
@@ -13,24 +13,15 @@ public TaskCompletionSourceAdapter(object taskCompletionSource)
 
         public Type ResponseType => taskCompletionSource.GetType().GenericTypeArguments[0];
 
-        public void SetResult(object result)
+        public void TrySetResult(object result)
```
Comment (andreasohlund) [inline src/NServiceBus.Callbacks/TaskCompletionSourceAdapter.cs]: It feels weird with a TryX that takes no `out` parameter and doesn't return `bool`?

Comment (SzymonPobiega) [inline src/NServiceBus.Callbacks/TaskCompletionSourceAdapter.cs]: Would would we do with the result if it is false (e.g. when a task has been cancelled)?

Comment (andreasohlund) [inline src/NServiceBus.Callbacks/TaskCompletionSourceAdapter.cs]: Not sure, I realize that was real reason I asked. Having a tryget where we don't check the result feels off?

Comment (SzymonPobiega) [inline src/NServiceBus.Callbacks/TaskCompletionSourceAdapter.cs]: OK. The reason I've put it here is that the previous code (`SetResult`) was blowing up with an exception but the acceptance test was declared `done` before the exception was caught. This attempts to fix the problem of cancelling the timeout. If the task is cancelled (`false` returned) I don't think we can do anything.

File: src/NServiceBus.Callbacks/TaskCompletionSourceAdapter.cs [hunk 7]
```diff
@@ -13,24 +13,15 @@ public TaskCompletionSourceAdapter(object taskCompletionSource)
...
         {
-            var method = taskCompletionSource.GetType().GetMethod("SetResult");
+            var method = taskCompletionSource.GetType().GetMethod("TrySetResult");
```
Comment (SimonCropp) [inline src/NServiceBus.Callbacks/TaskCompletionSourceAdapter.cs]: do we need to reflect this every time? perhaps a compiled delegate?

Comment (danielmarbach) [inline src/NServiceBus.Callbacks/TaskCompletionSourceAdapter.cs]: Currently yes. On the original PR we decided it is not worth it since callbacks tend to be used in "slow" integration scenarios. But happy to revise this decision

Comment (SzymonPobiega) [inline src/NServiceBus.Callbacks/TaskCompletionSourceAdapter.cs]: I'd say it is out of scope of this PR. To be honest, fixing the `SetResult` to `TrySetResult` is out of scope here also but decided to put it in because the test was not testing anything with the old code here.

Comment (andreasohlund) [inline src/NServiceBus.Callbacks/TaskCompletionSourceAdapter.cs]: Fair enough, I have nothing better to propose

**Comment by andreasohlund** [comment]
Fix the redundant `using` and perhaps a squash so that we can pull this?

File: src/NServiceBus.Callbacks/TaskCompletionSourceAdapter.cs (hunk 7 above)
Comment (SimonCropp) [inline src/NServiceBus.Callbacks/TaskCompletionSourceAdapter.cs]: Yeah ignore my reflection comment. As long as it has been considered

**Comment by SzymonPobiega** [comment]
Rebased, squashed and fixed the inspection.
